paf-mvp-operator-express/src/operator-node.ts

```diff
@@ -354,6 +372,87 @@ export class OperatorNode extends Node {
     setCookie(res, Cookies.test_3pc, toTest3pcCookie(test3pc), expirationDate, { domain: this.topLevelDomain });
   }
 
+  buildWriteIdsAndPreferencesSignatureHandler(isRedirect: boolean): RequestHandler {
+    return async (req: Request, res: Response, next: NextFunction) => {
+      const request = isRedirect
+        ? getPafDataFromQueryString<RedirectPostIdsPrefsRequest>(req)
+        : getPayload<PostIdsPrefsRequest>(req);
+      const validationResult = await this.validateWriteRequest(request, req);
+      if (validationResult.isValid) {
+        next();
+      } else {
+        const error: NodeError = {
+          type:
+            validationResult.errors[0] instanceof UnableToIdentifySignerError
+              ? NodeErrorType.UNKNOWN_SIGNER
+              : NodeErrorType.VERIFICATION_FAILED,
+          details: validationResult.errors[0].message,
+        };
+        if (isRedirect) {
+          this.redirectWithError(res, (request as RedirectPostIdsPrefsRequest).returnUrl, 403, error);
+        } else {
+          res.status(403);
+          res.json(error);
+        }
+        next(error);
```

Comment:
⚠️ **This point is the most important of my review.  I would like that we dig that together. Pair programming?**

For now, we have an Error Handler that only logs and we don't call `next(err)` in it. When we read [the official documentation](https://expressjs.com/en/guide/error-handling.html), the Error Handlers are about catching errors and updating the response accordingly. I think that there is something profoundly wrong with the way we handle errors in this app (not your fault).

I share with you a scenario: for some reason, a previous handler of `WriteIdsAndPreferencesSignatureHandler` or the method `getPadDataFronQueryString` throws an exception... Who catches it? Who does the redirect? 

I am not sure because the app is more complex than the simple example that I am going to share here but I think that nobody catches this error and the HTTP connection is kept alive for a while.

## Diagnose

Please see the following little example that explores how express handles errors:
```javascript
const express = require('express');

function redirect(req, res, next) {
    console.log(`Redirecting`);
    res.redirect(redirectURL);
    next(new Error('Redirection Error passed to next'));
}

function thrower(req, res, next) {
    console.log(`Throwing error`);
    throw new Error('Throwed error');
}

function logError(err, req, res, next) {
    console.log(`Logging Error WITHOUT next: ${err}`);
}

function logErrorNext(err, req, res, next) {
    console.log(`Logging Error WITH next: ${err}`);
    next(err);
}

function catcher(err, req, res, next) {
    console.log(`Catching error ${err}`);
    res.status(500).send(`Catched: ${err}`)
}

function log(req, res, next) {
    console.log(`\n-------Calling ${req.url}---------`);
    next();
}

function hello(req, res) {
    console.log(`Saying hello`);
    res.status(200);
    res.send('Hello World!');
}

const app = express()
const port = 3000
const redirectURL = 'http://localhost:3000';

app.use(log);
app.get('/', hello);

app.get('/no-catch',            thrower, redirect, logError);
app.get('/no-catch-with-next',  thrower, redirect, logErrorNext);

app.get('/catch',               thrower, redirect, logError,        catcher);
app.get('/catch-with-next',     thrower, redirect, logErrorNext,    catcher);

app.get('/catch-after-redirect', redirect, logErrorNext);

app.listen(port, () => {
  console.log(`Go to:`);
  
  console.log(`\thttp://localhost:${port}/catch`);
  console.log(`\thttp://localhost:${port}/catch-with-next`);
  
  console.log(`\thttp://localhost:${port}/no-catch`);
  console.log(`\thttp://localhost:${port}/no-catch-with-next`);

  console.log(`\thttp://localhost:${port}/catch-after-redirect`);
});
```

From what I see from the existing codebase, our endpoints are similar to `/no-catch`: if one of our handlers throws an exception, it bypasses the next handlers, then our custom Error Handler logs the error and waits for a while. Here is the console when we call `/no-catch`.

```
-------Calling /no-catch---------
Throwing error
Logging Error WITHOUT next: Error: Throwed error
```

Note also that adding `next(err)` in our current Logging Error Handler isn't enough because we would end up with unmanaged errors in the console even if (by luck) the service will respond correctly. See the output of `/catch-after-redirect`.

````
-------Calling /catch-after-redirect---------
Redirecting
Logging Error WITH next: Error: Redirection Error passed to next
Error: Redirection Error passed to next
    at redirect (/Users/r.lofaso/Dev/test-express/index.js:6:10)
    at Layer.handle [as handle_request] (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/layer.js:95:5)
    at /Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/index.js:284:15
    at Function.process_params (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/index.js:346:12)
    at next (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/index.js:280:10)
    at log (/Users/r.lofaso/Dev/test-express/index.js:30:5)
    at Layer.handle [as handle_request] (/Users/r.lofaso/Dev/test-express/node_modules/express/lib/router/layer.js:95:5)
-------Calling /---------
Saying hello

````
## How to fix

### For this review

For this review, I think that we should create dedicated Error Handlers that do the redirect anyway. Something like that:
```typescript
this.app.expressApp.get(
    redirectEndpoints.read,
    // ,,,
    this.redirectReadIdsAndPreferences,
    this.handleErrorsWithRedirectionToReferer(redirectEndpoints.read), // <----------------
    this.endSpan(redirectEndpoints.read)
  );


handleErrorsWithRedirectionToReferer(error: Error, re: Request, res: Response): void {
    const url = req.header('referer');
    this.logger.Info(`redirecting to ${url} ...`);
    httpRedirect(res, url);
};
```

I don't know if this handler matches all the use cases because I see that we try to extract the redirected URL from the query parameters. We can refine that together and explore other strategy for getting 'safely' a redirection URL but the idea of the Error Handler is here.

### For the whole app

For the whole app, I think that we should rework the generic Error Handler. It should write a custom response instead of logging and potentially call `endSpan()` itself. Moreover, we should stop to call `next(err)` when we define an error and use `res.status(...)` and `res.json(...)` ourselves.

Reply:
- RequestValidation handlers (Format + signature + permission) are not supposed to throw any error(unless it's a bug). For example `getPadDataFromQueryString` will never throw an error because it's called after requestSchema validation. But I totally agree with you to cover the cases of unexpected errors with the equivalent of a global try catch.
- One way of managing errors is to let the handlers throw errors and have the ErrorHandler catch them, update the response accordingly and send it. I think it's a good way of managing `Unexpected errors`.  What I did is that I let each RequestValidation Handler (Format + signature + permission) manage its own errors internally and kept the global ErrorHandler for unhandled errors for many reasons:  
              1- I think that, functionally speaking, managing validation errors is in the scope of a validation handler.
              2- The ErrorHandler code will be a big mess and will became unreadable if we let it manage all type of errors
              3- One limitation of express middlewares is that they can't share data natively (we still can do it the dirty way though). And sometimes having just the error object passed to the errorHandler is not enough. Let's take an example of a redirectRequest: Assume that an Error was thrown by some handler and it was catched by the errorHandler. Now, we should do a redirect but to which address? It depends on the step at which the error was thrown. If it was at requestSchemaValidation or redirectUrlValidation, it means that we need to redirect to referer because the redirectUrl is not available. Otherwise, we need to redo the parsing and extract redirectUrl in the errorHandler... 
              
For now, I will add the redirectErrorHandler you suggest to cover redirectRequests and I will rework the actual errorHandler later.

Reply:
> RequestValidation handlers (Format + signature + permission) are not supposed to throw any error(unless it's a bug). For example getPadDataFromQueryString will never throw an error because it's called after requestSchema validation. 

Yes, I think that the goal of this PR is to make the Redirection extremely resilient since we identified critical impacts when the redirection doesn't work properly. Therefore, I am concerned about the "errors due to bug" as much as the handled ones. Always be humble with our own code - issues happen. :-) 
 
> I let each RequestValidation Handler (Format + signature + permission) manage its own errors internally and kept the global ErrorHandler for unhandled errors for many reasons.

I can be aligned with this approach even if I don't think it is the most optimum for debugging. However, I don't think this is what exactly happens in the current code. See the line that I selected for this thread: `next(error)`. It means that, after *managing its own errors internally*, the handler still called the Error Handler similarly to thrown errors.

See the [official documentation](http://expressjs.com/en/guide/error-handling.html):

*If you pass anything to the next() function (except the string 'route'), Express regards the current request as being an error and will skip any remaining non-error handling routing and middleware functions.*

> 1- I think that, functionally speaking, managing validation errors is in the scope of a validation handler.

Agree considering the change about `next(error);`

>  The ErrorHandler code will be a big mess and will become unreadable if we let it manage all type of errors

Well, if you use your proposed approach, it is fine. If we use the error handler for every type of error (to be explored together in zoom maybe), it should not be a mess. Actually, it may even simplify the code.

Example that also includes the possibility to manage the case that your mentioned in your point 3:
```typescript
  interface NodeError
  {
    type: ErrorType,
    details: string,
    httpStatus: number;
    redirectUrl?: URL;
  }
  
  function buildReadIdsAndPreferencesSignatureHandler(isRedirect: boolean): RequestHandler {
    return async (req: Request, res: Response, next: NextFunction) => {
      const request = isRedirect
        ? getPafDataFromQueryString<RedirectGetIdsPrefsRequest>(req)
        : getPafDataFromQueryString<GetIdsPrefsRequest>(req);
      const validation = await this.validateReadRequest(request, req);
      next(errorIfInvalidSignature(validation));
    };
  }

  function errorIfInvalidSignature(req: Request, result: SignatureValidationResult) NodeError | undefined {
    if (signatureValidationResult.isValid) {
        return undefined;
    }

    const type =  result.errors[0] instanceof UnableToIdentifySignerError ? 
        NodeErrorType.UNKNOWN_SIGNER : 
        NodeErrorType.VERIFICATION_FAILED;
    const error: NodeError = {
        httpStatus: 403,
        type: type,
        details: signatureValidationResult.errors[0].message,
        redirectUrl: new URL(req.header('referer')),
    };
    return error;
  }

  /**
   * Transform all error to a Node Error (if needed).
   * Used before erroHandlerForRedirection or erroHandlerForRest.
   */
  function allErrorHandler(err: Error | string, req: Request, res: Response, next: Next) {
    if (err instanceof NodeError) {
        next(err);
    } else {
        next(nodeError(err));
    }
  }

  function erroHandlerForRedirection(err: NodeError, req: Request, res: Response, next: Next) {
    this.logger.Error(`Catched error for redirecting to ${url}: ${err}`);
    const url = err.redirectUrl ?? req.header('referer');
    try {
        const redirectURL = buildErrorRedirectUrl(new URL(url), err);
        httpRedirect(res, redirectURL.toString());
    } catch (e) {
        this.logger.Error(`Second error while redirecting to ${url}: ${e}`);
        httpRedirect(res, url);
    }
  }

  function erroHandlerForRest(err: NodeError, req: Request, res: Response, next: Next) {
    this.logger.Error(`Catched error for REST API: ${err}`);
    try {
        res.status(error.httpStatus);
        res.json(error);
    } catch (e) {
        this.logger.Error(`Second error for REST API to ${url}: ${e}`);
        res.status(500);
        res.text((isInProduction) ? "Unexpected error" : err.toString());
    }
  }

  function nodeError(err: Error | string): NodeError {
    const error: NodeError = {
            type: NodeErrorType.UNKNOWN,
            details: (isInProduction) ? "Unexpected error" : err.toString(),
            httpStatus: 500,
    }
    // Also add stack trace in the detail in we are not in production.
    return error;
  }
```